File: src/main/java/org/bazar/chat/app/api/message/MessageService.java
package org.bazar.chat.app.api.message;

import org.bazar.chat.app.api.message.dto.CreateMessageDto;
import org.bazar.chat.app.api.message.dto.GetMessagePageDto;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Сервис для работы со сценариями по сущности Сообщение
 */
public interface MessageService {
    /**
     * Получить сообщения чата с пагинацией
     *
     * @param chatId Идентификатор чата
     * @param pageable Информация по пагинации
     * @return Сообщения чата
     */
    GetMessagePageDto getChatMessages(Long chatId, Pageable pageable);

    /**
     * Создать сообщение
     *
     * @param dto DTO для создания сообщения
     */
    void createMessage(CreateMessageDto dto);

    /**
     * Удалить истекшие сообщения
     */
    void deleteExpiredMessages();

    /**
     * Удалить сообщения по идентификатору чата и идентификаторам сообщений
     *
     * @param chatId Идентификатор чата
     * @param messageIds Идентификаторы сообщений
     */
    void deleteMessages(Long chatId, List<Long> messageIds);

    /**
     * Обновить содержимое сообщения
     *
     * @param updateMessageDto DTO для обновления содержимого сообщения
     */
    void updateMessageContent(UpdateMessageDto updateMessageDto);
}

File: src/main/kotlin/org/bazar/chat/service/MessageService.kt
package org.bazar.chat.service

import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.transaction.Transactional
import org.bazar.chat.model.CreateMessageRequest
import org.bazar.chat.persistence.entity.Chat
import org.bazar.chat.persistence.entity.Message
import org.bazar.chat.persistence.repository.MessageRepository
import org.bazar.chat.utils.extensions.getUserIdFromSecurityContext
import org.bazar.chat.utils.extensions.toMessagePageResponse
import org.bazar.chat.utils.extensions.toMessageResponse
import org.springframework.data.domain.Pageable
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.stereotype.Service
import kotlin.math.log

@Service
@Transactional
class MessageService(
    private val messageRepository: MessageRepository,
    private val chatService: ChatService,
    private val webSocketTemplate: SimpMessagingTemplate
) {

    private val logger = KotlinLogging.logger { }

    fun createMessage(chatId: Long, createMessageRequest: CreateMessageRequest) {
        val message =
            messageRepository.save(
                Message(
                    Chat(id = chatId, spaceId = chatService.getChatById(chatId)),
                    createMessageRequest.content,
                    getUserIdFromSecurityContext()
                )
            )
        val destination = "/topic/chat/${message.chat.id}"
        logger.debug { "SENDING TO $destination" }
        webSocketTemplate.convertAndSend(destination, message.toMessageResponse())
    }

    fun getMessages(chatId: Long, pageable: Pageable) =
        messageRepository.findAllByChatIdOrderByCreatedAtDesc(chatId, pageable).toMessagePageResponse()

}